Support setting the button aspect ratio

marienz opened the pull request:
This looks better when the aspect ratio of the grid is very different from the aspect ratio of the monitor (resulting in very tall or very wide buttons). This is not always avoidable by using a button layout that fits the monitor (for example, when using multiple monitors with very different aspect ratios).

On one of my monitors without this patch:

<img width="2560" height="2880" alt="tall" src="https://github.com/user-attachments/assets/05450879-e3a7-450a-aa77-7ed758c8321e" />

With this patch and `--button-aspect-ratio 1`:

<img width="2560" height="2880" alt="fixed" src="https://github.com/user-attachments/assets/fe2fa890-67e9-431d-92a0-045cb859554c" />

...and I can't just use a more square button arrangement because I have a "normal" 16:9 monitor sitting next to it, where this 2x4 grid makes sense.

This is a draft because of two issues:

- The version info can end up far removed from the buttons:

<img width="2560" height="2880" alt="fixed-with-version" src="https://github.com/user-attachments/assets/e99d7407-7d48-49ff-995a-cee6ca64d7e8" />

I haven't figured out a way of fixing this yet (but I also haven't wrapped my head around how Gtk layout actually works yet, so it might well be easy, hints appreciated!). If I don't allow the grid and/or buttons to consume extra space, everything shrinks to the size of the icons in the buttons... (the buttons get much smaller).

- Specifying the ratio as a float is maybe not the best (it should maybe be a fraction like how buttons-per-row works?)

Any feedback much appreciated.

AMNatty commented:
Good idea, I like it.

Web CSS has an option for setting an aspect ratio so I'd prefer the syntax to follow that. In a different project I use an actual CSS parser, however I think that would be way overkill here, so I think accepting both a ratio and a floating point value would be nice.

As for the version info, I'll try to look into it. I think there might be some tech debt related to the side margins, so it might be a good idea to revisit for how the box interacts with the margins and how it grows maybe?

Last point (for reference), the option should be documented in the man docs, completions, and in the project README. Other than that, so far so good.

marienz commented:
> accepting both a ratio and a floating point value would be nice

> option should be documented in the man docs, completions, and in the project README

Thanks, noted, I'll get those implemented once this actually works...

Assuming my (rudimentary) understanding of how Gtk layout works is correct, GtkAspectFrame doesn't quite do what I want :(

What I was trying to do:

- Pack an AspectFrame and the version-info label into the outer vertical box (`container_box`)
- Horizontally expand the AspectFrame (so it fills the width of the screen)
- Make the AspectFrame take up height based on that width and its set aspect ratio
- Vertically pack the AspectFrame and label together in `container_box`

Given the description of [height-for-width geometry management](https://docs.gtk.org/gtk4/class.Widget.html) in the manual this seemed like it should work, but there is a problem: AspectFrame [does not look at ratio in `gtk_aspect_frame_measure`](https://gitlab.gnome.org/GNOME/gtk/-/blob/a885dda8c9daacba747a1dd9f81dba0e407bf1f9/gtk/gtkaspectframe.c#L570-597) (which drives geometry management), it just forwards it to the child widget. (It only looks at ratio in functions called from `gtk_aspect_frame_size_allocate`, which if I understand correctly runs *after* its container has decided its size).

Which means that without vexpand on the AspectFrame, when container_box checks for the AspectFrame's desired height, it gets a much smaller value than we want. And when the AspectFrame is then asked to use (allocate) that smaller height, AspectFrame shrinks the buttons to maintain our set aspect ratio.

I haven't spotted an easy way of overriding this behavior (short of "custom widget", which is further complicated by AspectFrame not being subclassable... possible, but more code than this feature warrants).

We can kind of work around this by putting the buttons and the label both inside the AspectFrame: since the label isn't all that tall this gets us mostly the correct layout.

Remaining problem: although changing `--button-aspect-ratio` changes the button's aspect ratio, the actual aspect ratio we end up with doesn't match what we set (I haven't tracked down why yet, it's not just the version-info label skewing things).

But at least it looks *about* right:

<img width="2560" height="2880" alt="mostly-fixed" src="https://github.com/user-attachments/assets/b9f61d7d-72d6-4c73-9223-c044613c23eb" />

There's one more thing I want to try: [ConstraintLayout](https://docs.gtk.org/gtk4/class.ConstraintLayout.html), which sounds like it can replace both the outer box and AspectFrame... Complication there might be to make it behave correctly both if the screen is too tall and if it's too narrow (leaving horizontal or vertical gaps as required), but if that works it might be simpler than dealing with Box and AspectFrame interacting.

marienz commented:
> ConstraintLayout

...is a bit fiddly but I think I got it working.

It's quite verbose to use, adding >100 lines of constraint definitions. Being able to use [add_constraints_from_description](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.ConstraintLayout.html#method.add_constraints_from_description) would be quite a bit more ergonomic, but because of what looks like an oversight in the Rust bindings it requires actual widgets, not `GtkConstraintTarget`s, which means we can't use it with guides (which we need).

If it sounds ok, I think I'll finish up the AspectFrame approach (the one currently in this PR, which works for *approximately* setting the desired aspect ratio) and in parallel try to get `add_constraints_from_description` fixed upstream, then send a new PR to switch to that once released (which I'm assuming will take a while)..

marienz committed:
tweak: replace insert_child_after with append()

This does not require keeping track of what the preceding widget was.

marienz committed:
Support setting the button aspect ratio

This looks better when the aspect ratio of the grid is very different from the aspect ratio of the monitor (resulting in
very tall or very wide buttons). This is not always avoidable by using a button layout that fits the monitor (for
example, when using multiple monitors with very different aspect ratios).

The AspectBox contains both the grid and version_info, because we need to set it to expand vertically to get the buttons
to occupy the screen (this is necessary because AspectBox does not take its fixed aspect ratio into account in its
"measure" implementation, so the vertical Box it is in does not give it enough height to render the buttons at full
width). And with the AspectBox expanding vertically, if the version_info is not inside it it can end up far below the
buttons.

This mostly works, but the resulting aspect ratio only approximately matches the desired one (and it's further off than
just the label can explain, reasons unclear...).

AMNatty commented:
See the commit https://github.com/AMNatty/wleave/commit/4f514b4c151b4fc485809727a4bc099c66f2848c  for the idea. I think using a custom layout manager entirely may allow some special effects if desired, however I am not yet sure how to implement them in a backward-compatible manner.

AMNatty commented:
The idea is that it will allow us to dynamically compute the best possible grid on the fly, without having to specify the buttons per row option. I think it'd be nice to make it automatic by default, with the old behavior becoming opt-in.
